packages/firestore/src/local/local_documents_view.ts

```diff
@@ -532,36 +554,153 @@ export class LocalDocumentsView {
-            mutationApplyToLocalView(
-              overlay.mutation,
-              document,
-              FieldMask.empty(),
-              Timestamp.now()
+  private getDocumentsMatchingPipeline(
+    txn: PersistenceTransaction,
+    pipeline: CorePipeline,
+    offset: IndexOffset,
+    context?: QueryContext
+  ): PersistencePromise<DocumentMap> {
+    if (getPipelineSourceType(pipeline) === 'collection_group') {
+      const collectionId = getPipelineCollectionGroup(pipeline)!;
+      let results = documentMap();
+      return this.indexManager
+        .getCollectionParents(txn, collectionId)
+        .next(parents => {
+          // Perform a collection query against each parent that contains the
+          // collectionId and aggregate the results.
+          return PersistencePromise.forEach(parents, (parent: ResourcePath) => {
+            const collectionPipeline = asCollectionPipelineAtPath(
+              pipeline,
+              parent.child(collectionId)
             );
+            return this.getDocumentsMatchingPipeline(
+              txn,
+              collectionPipeline,
+              offset,
+              context
+            ).next(r => {
+              r.forEach((key, doc) => {
+                results = results.insert(key, doc);
+              });
+            });
+          }).next(() => results);
+        });
+    } else {
+      // Query the remote documents and overlay mutations.
+      let overlays: OverlayMap;
+      return this.getOverlaysForPipeline(txn, pipeline, offset.largestBatchId)
+        .next(result => {
+          overlays = result;
+          switch (getPipelineSourceType(pipeline)) {
+            case 'collection':
+              return this.remoteDocumentCache.getDocumentsMatchingQuery(
+                txn,
+                pipeline,
+                offset,
+                overlays,
+                context
+              );
+            case 'documents':
+              let keys = documentKeySet();
+              for (const key of getPipelineDocuments(pipeline)!) {
+                keys = keys.add(DocumentKey.fromPath(key));
+              }
+              return this.remoteDocumentCache.getEntries(txn, keys);
+            case 'database':
+              return this.remoteDocumentCache.getAllEntries(txn);
+            default:
+              throw new FirestoreError(
+                'invalid-argument',
+                `Invalid pipeline source to execute offline: ${canonifyPipeline(
+                  pipeline
+                )}`
+              );
           }
-          // Finally, insert the documents that still match the query
-          if (queryMatches(query, document)) {
-            results = results.insert(key, document);
-          }
+        })
+        .next(remoteDocuments => {
+          return this.retrieveMatchingLocalDocuments(
+            overlays,
+            remoteDocuments,
+            doc => pipelineMatches(pipeline, doc as MutableDocument)
```

Comment:
It isn't clear to me, but do we remove offsets and limits from the `pipeline` argument? If not, could those cause false negatives?

Reply:
We do not support offset in offline eval.

Limit will work as intended here, all docs matching the pipeline without limit will return, and limit will be applied later at view layer. 

We had discussion on whether we should move that logic into here, but no action had been taken so far.